model/src/w3src4md.F90

```diff
@@ -1791,31 +1894,69 @@ SUBROUTINE CALC_USTAR(WINDSPEED,TAUW,USTAR,Z0,CHARN)
     !
     ! 10. Source code :
     !-----------------------------------------------------------------------------!
-    USE CONSTANTS, ONLY: GRAV, KAPPA
-    USE W3GDATMD,  ONLY: ZZWND, AALPHA
+    USE CONSTANTS, ONLY: GRAV, KAPPA, NU_AIR
+    USE W3GDATMD,  ONLY: ZZWND, AALPHA, ZZ0MAX, SINTAILPAR
 #ifdef W3_T
     USE W3ODATMD, ONLY: NDST
 #endif
     IMPLICIT NONE
     REAL, intent(in) :: WINDSPEED,TAUW
     REAL, intent(out) :: USTAR, Z0, CHARN
     ! local variables
-    REAL SQRTCDM1
-    REAL XI,DELI1,DELI2,XJ,delj1,delj2
-    REAL TAUW_LOCAL
-    INTEGER IND,J
-    !
-    TAUW_LOCAL=MAX(MIN(TAUW,TAUWMAX),0.)
-    XI      = SQRT(TAUW_LOCAL)/DELTAUW
-    IND     = MIN ( ITAUMAX-1, INT(XI)) ! index for stress table
-    DELI1   = MIN(1.,XI - REAL(IND))  !interpolation coefficient for stress table
-    DELI2   = 1. - DELI1
-    XJ      = WINDSPEED/DELU
-    J       = MIN ( JUMAX-1, INT(XJ) )
-    DELJ1   = MIN(1.,XJ - REAL(J))
-    DELJ2   = 1. - DELJ1
-    USTAR=(TAUT(IND,J)*DELI2+TAUT(IND+1,J  )*DELI1)*DELJ2 &
-         + (TAUT(IND,J+1)*DELI2+TAUT(IND+1,J+1)*DELI1)*DELJ1
+    REAL             :: SQRTCDM1
+    REAL             :: XI,DELI1,DELI2,XJ,delj1,delj2  ! used for table version
+    INTEGER          :: IND,J
+    REAL             :: TAUW_LOCAL
+    REAL             :: TAUOLD,CDRAG,WCD,USTOLD,X,UST,ZZ0,ZNU,ZZ00,F,DELF
+    INTEGER, PARAMETER      :: NITER=10
+    REAL   , PARAMETER      :: XM=0.50, EPS1=0.00001
+    INTEGER                 :: ITER
+    !     VARIABLE.   TYPE.     PURPOSE.
+    !      *XM*        REAL      POWER OF TAUW/TAU IN ROUGHNESS LENGTH.
+    !      *XNU*       REAL      KINEMATIC VISCOSITY OF AIR.
+    !      *NITER*     INTEGER   NUMBER OF ITERATIONS TO OBTAIN TOTAL STRESS
+    !      *EPS1*      REAL      SMALL NUMBER TO MAKE SURE THAT A SOLUTION
+    !                            IS OBTAINED IN ITERATION WITH TAU>TAUW.
+
+    !
+    IF (SINTAILPAR(1).GT.0.5) THEN
```

Comment:
@StevePny This may impact the computation of ustar, z0 and charnock. I think worth testing before making a new build with this version.